### PR TITLE
Checkstyle reporter

### PR DESCRIPTION
**Pull request by jonathan-dev**
Added the possibility to export the linting results as checkstyle xml by using the `--checkstyle` flag on the lint subcommand.
The `--teamcity` flag takes precidence over `--checkstyle`.

The xml contents are written to stdout and thus can be forwarded to a file without containing the logs which are send to stderr (using `1>`).

This feature was implemented in order to work with the [jenkins warnings plugin](https://plugins.jenkins.io/warnings-ng/)

### CODE REVIEW

**Comment by jonathan-dev** [comment]
https://github.com/cloudflare/pint/issues/606

**Comment by prymitive** [comment]
> Added the possibility to export the linting results as checkstyle xml by using the `--checkstyle` flag on the lint subcommand.

I, in general, don't want to add code that integrates with something that we don't run ourselves, as it just makes it impossible to refactor that code in the future.
I'm fine with adding a new reporter like this, as it's pretty much self contained, but only if it has good test coverage, which this one does.

> The `--teamcity` flag takes precidence over `--checkstyle`.
> The xml contents are written to stdout and thus can be forwarded to a file without containing the logs which are send to stderr (using `1>`).

You need a file at the end, so why not make it `--checkstyle=path.xml` flag and leave `--teamcity` as is? In theory there could be any number of report files produced in a variety of formats but sending it all to stdout makes it impossible to enable multiple such reporters at once.

**Comment by prymitive** [comment]
An e2e test like https://github.com/cloudflare/pint/blob/main/cmd/pint/tests/0158_lint_teamcity.txt would be great.

**Comment by jonathan-dev** [comment]
> > Added the possibility to export the linting results as checkstyle xml by using the `--checkstyle` flag on the lint subcommand.
> 
> I, in general, don't want to add code that integrates with something that we don't run ourselves, as it just makes it impossible to refactor that code in the future. I'm fine with adding a new reporter like this, as it's pretty much self contained, but only if it has good test coverage, which this one does.
> 
> > The `--teamcity` flag takes precidence over `--checkstyle`.
> > The xml contents are written to stdout and thus can be forwarded to a file without containing the logs which are send to stderr (using `1>`).
> 
> You need a file at the end, so why not make it `--checkstyle=path.xml` flag and leave `--teamcity` as is? In theory there could be any number of report files produced in a variety of formats but sending it all to stdout makes it impossible to enable multiple such reporters at once.

I change `--checkstyle` to be a string flag

**Comment by jonathan-dev** [comment]
> An e2e test like https://github.com/cloudflare/pint/blob/main/cmd/pint/tests/0158_lint_teamcity.txt would be great.

I created a test I am not comparing stderr because it was giving me trouble because of https://github.com/cloudflare/pint/issues/20 and i don't think it is relevant.

File: cmd/pint/ci.go [hunk 1]
```diff
@@ -118,7 +125,13 @@ func actionCI(c *cli.Context) error {
 	}
 
 	reps := []reporter.Reporter{}
-
+	if c.String(checkStyleFlag) != "" {
+		f, fileErr := os.Create(c.String(checkStyleFlag))
```
Comment (prymitive) [inline cmd/pint/ci.go:129]: Don't you need to close `f` before pint exits?

File: cmd/pint/lint.go [hunk 2]
```diff
@@ -100,16 +106,26 @@ func actionLint(c *cli.Context) error {
...
 	} else {
-		r = reporter.NewConsoleReporter(os.Stderr, minSeverity)
+		reps = append(reps, reporter.NewConsoleReporter(os.Stderr, minSeverity))
 	}
 
-	err = r.Submit(summary)
-	if err != nil {
-		return err
+	if c.String(checkStyleFlag) != "" {
+		f, fileErr := os.Create(c.String(checkStyleFlag))
```
Comment (prymitive) [inline cmd/pint/lint.go:117]: Same here
`Close()` might error out so we should check for that

File: internal/reporter/checkstyle.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	for _, report := range summary.reports {
+		x[report.Path.Name] = append(x[report.Path.Name], report)
+	}
+	return x
+}
+
+func (cs CheckStyleReporter) Submit(summary Summary) error {
+	dirs := sortByFile(summary)
+	var buf strings.Builder
+	buf.WriteString("<?xml version='1.0' encoding='UTF-8'?>\n")
```
Comment (prymitive) [inline internal/reporter/checkstyle.go:34]: `encoding/xml` can generate the entire XML file for you, that should be safer & more correct than manually crafting it.

File: internal/reporter/checkstyle.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+	return CheckStyleReporter{
+		output: output,
+	}
+}
+
+type CheckStyleReporter struct {
+	output io.Writer
+}
+
+type Dirs map[string][]Report
```
Comment (prymitive) [inline internal/reporter/checkstyle.go:21]: this doesn't have to be exported, does it?

File: internal/reporter/checkstyle.go (hunk 3 above)
Comment (jonathan-dev) [reply internal/reporter/checkstyle.go:34]: Sorry here I am not sure what you are referring to exactly. I see I could be using `xml.Header`.
Do you mean by having a custom struct that marshals to `<checkstyle version='4.3'>` making dirs marshal to `<file name=....` and report to <error line=.../>?

**Pull request merged**